### ISSUE TITLE
New security rules (NULL check)

### ISSUE THREAD

**Issue by icy17** [libevent/libevent#1692]
Hi, I've observed that some security rules are absent from the documentation.

For the security rule "parameters must not be NULL," its omission in the documentation can lead to significant security issues.

Firstly, while best programming practices dictate that developers should avoid passing NULL parameters to an API, this guideline may not be obvious to less experienced developers, increasing the risk of overlooking necessary security checks.

Secondly, many APIs receive parameters from other functions or APIs, which might not be adequately checked for NULL, posing a risk even for experienced developers.

Therefore, it is crucial to explicitly mention in the documentation that NULL checks must be conducted on specified parameters before an API is used. Including these security rules will help developers identify potential security issues more effectively during testing.

I will provide a list of APIs requiring checks along with their corresponding parameters：
| API | parameter index (begin with 1) | 
|:----------:|:----------:|
| event_del | 1 |
| evhttp_add_header | 1 |
| bufferevent_enable | 1 |
| event_base_dispatch | 1 |
| bufferevent_get_output | 1 |
| bufferevent_setcb | 1 |
| bufferevent_get_input | 1 |
| bufferevent_disable | 1 |
| bufferevent_free | 1 |
| evutil_inet_pton | 2 |
| evutil_ascii_strcasecmp | 1 |
| evutil_ascii_strcasecmp | 2 |
| evhttp_request_get_connection | 1 |
| evhttp_uri_free | 1 |
| event_initialized | 1 |
| event_debug_unassign | 1 |
| evutil_inet_ntop | 2 |
| bufferevent_socket_new | 1 |
| bufferevent_getfd | 1 |
| bufferevent_setwatermark | 1 |
| evdns_base_nameserver_add | 1 |
| evdns_base_nameserver_ip_add | 1 |
| evdns_base_set_option | 1 |
| evdns_base_resolv_conf_parse | 1 |
| evhttp_connection_set_timeout | 1 |
| evhttp_free | 1 |

**Comment by azat** [comment]
libevent historically do not check for NULL, and not all functions has return value to add a NULL check (how to propogate the error to user?)

So what I'm suggesting is to update documentation instead.

**Comment by icy17** [comment]
> libevent historically do not check for NULL, and not all functions has return value to add a NULL check (how to propogate the error to user?)
> 
> So what I'm suggesting is to update documentation instead.

I agree. I found that when I reported this issue to other developers, they suggested it would be more convenient to add a note in the documentation for APIs that return NULL (likeevent_base_new), rather than adding notes to the APIs that use the return value. However, sometimes when the caller uses an API like event_base_new, the caller might not need to check the return value because a subsequent call to event_new will handle this. I'm not sure which approach is better for libevent. Additionally, I'm not sure if I can open PRs and add patches to the documentation?

**Comment by azat** [comment]
>And I'm not sure if I can open PRs and add patches to the documentation?

Sure!

We use doxygen to generate documentation from sources, so just find relevant functions and add a note for them.
